### PR TITLE
New Resource: `azurerm_app_service`

### LINKED ISSUE #35: Azure web app deployment
_This issue was originally opened by @dpnl87 as hashicorp/terraform#11154. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi there,

For a new project I need to depoy Azure web apps (https://azure.microsoft.com/en-us/services/app-service/web/?b=16.52).

It looks like the azurerm provisioner has no resources to do this. Would it be hard to add this functionality?

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR extends the work done by @lstolyarov in #2 to support provisioning App Services within an App Service Plan.

There's a couple of limitations with this PR due to API issues (which I'll open Rest API Specs issues about shortly):
 - [it's currently not possible to update three fields: `client_affinity_enabled`, `enabled` and `tags` - as the Update endpoint has issues - such I've made them `ForceNew` for the moment](https://github.com/Azure/azure-rest-api-specs/issues/1697)
 - [it's currently only possible to provision Windows App Services, so I've removed references to Linux App Services for the moment (until the correct API fields are returned)](https://github.com/Azure/azure-rest-api-specs/issues/1698)

Fixes #35

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Tests pass:

screen shot 2017-09-18 at 15 08 27

File: azurerm/resource_arm_app_service.go [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+		properties := web.ConnectionStringDictionary{
+			Properties: connectionStrings,
+		}
+
+		_, err := client.UpdateConnectionStrings(resGroup, name, properties)
+		if err != nil {
+			return fmt.Errorf("Error updating Connection Strings for App Service %q: %+v", name, err)
+		}
+	}
+
```
Comment (stack72) [inline azurerm/resource_arm_app_service.go]: do any of these need to check to make sure it's Not a new resource?

File: azurerm/resource_arm_app_service.go [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["sites"]
+
+	resp, err := client.Get(resGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			// TODO: debug logging
```
Comment (stack72) [inline azurerm/resource_arm_app_service.go]: if it's a to-do, then.... ;)

File: website/docs/r/app_service.html.markdown [hunk 3]
```diff
@@ -0,0 +1,116 @@
+---
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_app_service"
+sidebar_current: "docs-azurerm-resource-app-service-x"
+description: |-
+  Manages an App Service (within an App Service Plan).
+
+---
+
+# azurerm\_app\_service
```
Comment (stack72) [inline website/docs/r/app_service.html.markdown]: we don't need the \ - this is a fallacy about markdown :)

File: website/docs/r/app_service.html.markdown [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+Manages an App Service (within an App Service Plan).
+
+## Example Usage
```
Comment (stack72) [inline website/docs/r/app_service.html.markdown]: I would really like to see 1 or 2 other example usages here - something maybe a little more complex with a connection_string struct maybe?

File: website/docs/r/app_service.html.markdown [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
+
+* `app_service_plan_id` - (Required) The ID of the App Service Plan within which to create this App Service. Changing this forces a new resource to be created.
+
+* `app_settings` - (Optional) A key-value pair of App Settings.
+
+* `connection_string` - (Optional) An `connection_string` block as defined below.
+
+* `client_affinity_enabled` - (Optional) TODO: COMPLETE ME. Changing this forces a new resource to be created.
```
Comment (stack72) [inline website/docs/r/app_service.html.markdown]: */me whistles....*

**Review by stack72 — COMMENTED** [review]
@tombuildsstuff this is looking awesome - I left a comments inline for you to think of before it get's merged

One thing I am really not particularly a fan of is these types of validations - just for info, Tom and I spoke on the phone about this so he understands my concerns and I understand his reasoning - it's just nice to let it be known that we have spoken

File: azurerm/resource_arm_app_service.go [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+	log.Printf("[INFO] preparing arguments for AzureRM App Service creation.")
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	location := d.Get("location").(string)
+	appServicePlanId := d.Get("app_service_plan_id").(string)
+	enabled := d.Get("enabled").(bool)
+	tags := d.Get("tags").(map[string]interface{})
+
+	siteConfig := expandAppServiceSiteConfig(d)
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: nope, the schema validation should handle this (so I don't think there's anything needed here)

File: azurerm/resource_arm_app_service.go [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+	appSettings := expandAppServiceAppSettings(d)
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: (as above, there's no error thrown here - so not at this time)

File: azurerm/resource_arm_app_service.go [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+	if v, ok := d.GetOk("client_affinity_enabled"); ok {
+		enabled := v.(bool)
+		siteEnvelope.SiteProperties.ClientAffinityEnabled = utils.Bool(enabled)
+	}
+
+	// NOTE: these seem like sensible defaults, in lieu of any better documentation.
+	skipDNSRegistration := false
+	forceDNSRegistration := false
+	skipCustomDomainVerification := true
+	ttlInSeconds := "60"
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: ¯\_(ツ)_/¯

File: azurerm/resource_arm_app_service.go [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+
+						"dotnet_framework_version": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "v4.0",
+							ValidateFunc: validation.StringInSlice([]string{
```
[the remark this answers is not in the document]
Comment (mbfrahry) [reply azurerm/resource_arm_app_service.go]: This bit us with another provider so I'd agree with Stack here

File: website/docs/r/app_service.html.markdown [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+* `enabled` - (Optional) TODO: COMPLETE ME. Changing this forces a new resource to be created.
```
Comment (mbfrahry) [inline website/docs/r/app_service.html.markdown]: Here too

**Review by mbfrahry — COMMENTED** [review]
Agreeing with Stack on quite a few things but other than that it looks really good.

File: azurerm/resource_arm_app_service.go [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+		},
+	}
+
+	if v, ok := d.GetOk("client_affinity_enabled"); ok {
+		enabled := v.(bool)
+		siteEnvelope.SiteProperties.ClientAffinityEnabled = utils.Bool(enabled)
+	}
+
+	// NOTE: these seem like sensible defaults, in lieu of any better documentation.
+	skipDNSRegistration := false
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: so the same API is used by the Azure CLI / PowerShell - which is why these options exist.

At this point in time we're not supporting using custom domain names - as such there's no value in exposing these fields, or in other cases (such as the delete options) - would delete parent resources (which should be being deleted by Terraform) -hence the sensible defaults

File: azurerm/resource_arm_app_service.go (hunk 1 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: nope, they're all applicable here

File: azurerm/resource_arm_app_service.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: updated

File: azurerm/resource_arm_app_service.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: Given this refers to versions of .net, I don't think this is likely to change often (if at all, now .net core is out), as such I think the validation is safe enough. Submitting `v2.0` to the API actually provisions the latest version of .net 2.0 (which is currently `.net 3.5.x`) and submitting `v4.0` uses the latest version of 4.x (currently `.net 4.7`).

Given this (and any new version of .net being supported being likely to be a big announcement) - I'm going to suggest we leave this in for the moment, until it's proven to be a pain

File: azurerm/resource_arm_app_service.go [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+	_, createErr := client.CreateOrUpdate(resGroup, name, siteEnvelope, &skipDNSRegistration, &skipCustomDomainVerification, &forceDNSRegistration, ttlInSeconds, make(chan struct{}))
+	err := <-createErr
+	if err != nil {
+		return err
+	}
+
+	settings := web.StringDictionary{
+		Properties: appSettings,
+	}
+	_, err = client.UpdateApplicationSettings(resGroup, name, settings)
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: updated - tests still pass

**Comment by tombuildsstuff** [comment]
Updated - tests still pass:

screen shot 2017-09-19 at 17 34 03

File: website/docs/r/app_service.html.markdown (hunk 10 above)
Comment (tombuildsstuff) [reply website/docs/r/app_service.html.markdown]: updated

File: website/docs/r/app_service.html.markdown (hunk 3 above)
Comment (tombuildsstuff) [reply website/docs/r/app_service.html.markdown]: updated

File: website/docs/r/app_service.html.markdown (hunk 4 above)
Comment (tombuildsstuff) [reply website/docs/r/app_service.html.markdown]: updated this to be a `.net 4.x` example with a sql connection string, and added a new java example

File: website/docs/r/app_service.html.markdown (hunk 5 above)
Comment (tombuildsstuff) [reply website/docs/r/app_service.html.markdown]: added a Java

**Review by mbfrahry — APPROVED** [review]
LGTM!

File: azurerm/resource_arm_app_service.go [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+				Optional: true,
+				Computed: true,
+				ForceNew: true, // due to a bug in the Azure API :(
+			},
+
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true, // due to a bug in the Azure API :(
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_app_service.go]: Updated to link to https://github.com/Azure/azure-rest-api-specs/issues/1697

**Comment by elemantalcode** [comment]
If I could add more +1 or +<3 I would... thank you!!!

**Comment by pixelicous** [comment]
@tombuildsstuff can we add support for msi, custom domain name, support for https only etc?

Currently i need to do that through arm template.. but i get "ttlInSeconds object is not present in the request body"

**Comment by tombuildsstuff** [comment]
@pixelicous 

> can we add support for msi, custom domain name, support for https only etc?

Issue #756 is tracking the feature request for these items - which we're hoping to look into in the near future.

> "ttlInSeconds object is not present in the request body

Unfortunately this is a bug in the Swagger/API which has been raised with Microsoft/Azure here: https://github.com/Azure/azure-rest-api-specs/issues/1697

**Comment by pixelicous** [comment]
@tombuildsstuff thank you! i actually manage to get around the msi problem with an arm template, i had a problem getting the identity.principalid though, that made me get stuck because spns cannot be created using arm template, and using powershell with external provider is just ugly code, you need to export profile of spn, and use it, and remove it.. becomes very ugly.

On another note, Tom i was wondering if you can shade some light, i have been in a couple of hashicorp/ms conferences and they speak about the great cooperation between the companies. now this has been said for months on end now, still, not much change on the azurerm provider.. it seems like you are the only one working on it, and in terms of module registries, well aws has around 200 while azure has around 20, so whats the gig.. what is the workforce behind this provider.. this is very lagging behind to the point that i am thinking of finding another tool to get the job done

thanks man, hopefully you could provide some answer for this, and sorry for hijacking the reply 😋

**Pull request merged**